### PR TITLE
🔘 parser for bindings text

### LINKED ISSUE #12: 📜 write a parser and builder for the bindings language

### PR DESCRIPTION
**Pull request by zenkyuv**

### CODE REVIEW

File: s/elements/context/parts/parse-bindings-text.test.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+					👼 alpha
+					👼 bravo
+				`)
+				expect(comments.length).equals(1)
+				const [a, b] = comments
+				expect(a.trim()).equals("alpha")
+				expect(b.trim()).equals("bravo")
+			},
+		}
+	},
```
Comment (chase-moskal) [inline s/elements/context/parts/parse-bindings-text.test.ts:109]: :warning: all of these tests related to comments are failing.

- we want to change the way the parser handles comments
- basically, we want to allow comments to contain any whitespace, including everything up to the next emoji, or end of file

these failing tests are specifications, and also requirements, for how the parser needs to behave.

File: s/elements/context/parts/parse-bindings-text.test.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+		},
+
+		"duplicate action in channel": async() => {
+			expect(() => parseBindingsText(`
+				🕹️ benevolent :: alpha
+				🕹️ benevolent :: bravo
+			`)).throws()
+		},
+	},
+}
```
Comment (chase-moskal) [inline s/elements/context/parts/parse-bindings-text.test.ts:128]: :warning: these tests, ironically about failure, are failing.

we want the parser to throw an error, in each of these three cases.

File: s/types.ts [hunk 3]
```diff
@@ -35,19 +35,23 @@ export namespace Nub {
...
+	"🖱": ActionBinds
 
 	/** vector2 action bindings */
-	"🕹️": {[action: string]: string[]}
+	"🕹️": ActionBinds
 
 	/** key action bindings */
-	"*️⃣": {[action: string]: string[]}
+	"*️⃣": ActionBinds
 }
```
Comment (chase-moskal) [inline s/types.ts:55]: i've taken to calling the emojis "channels", which have "binds"..

let's set some terminology for the action binds syntax:

```
🕹️ look :: alpha bravo
```

`🕹️` — channel  
`look` — action  
`alpha` — item  
`bravo` — item  

"bind" — a record of the relationship between an "action" and any number of "items"

:thinking: note: i did use the term `name` a lot, where now i would prefer `item`.. we'll probably refactor it to `item` sooner or later

**Review by chase-moskal — CHANGES_REQUESTED** [review]
hey @zenkyuv, great work on this so far!

- i pushed a commit that adds a test suite i wrote for the parser
- your parser is passing 14/20 tests
- 6 tests are failing, because they represent more specific requirements that we need to satisfy

let's adjust the parser and satisfy the tests before merging this branch

File: s/elements/bindings-editor/element.ts [hunk 4]
```diff
@@ -36,7 +36,12 @@ export const NubBindingsEditor = element({
 		</div>
 
 		${showTextEditor
-			? TextEditorPanelView({bindings})
+			? TextEditorPanelView({
+				bindings,
+				onClickSave(draft) {
+					context.updateBindings(draft)
+				},
+			})
```
Comment (chase-moskal) [inline s/elements/bindings-editor/element.ts:44]: here i've implemented the text editor saving changes

File: s/elements/bindings-editor/views/text-editor-panel.ts [hunk 5]
```diff
@@ -2,22 +2,53 @@
...
 export const TextEditorPanelView = view(use => ({
 		bindings,
+		onClickSave,
 	}: {
 		bindings: Bindings
+		onClickSave: (draft: Bindings) => void
 	}) => {
 
+	const [error, setError] = use.state<undefined | Error>(undefined)
+	const [draft, setDraft] = use.state<undefined | Bindings>(undefined)
```
Comment (chase-moskal) [inline s/elements/bindings-editor/views/text-editor-panel.ts:18]: the text editor now can be in an `error` state, and also keeps track of whether the user has created a `draft` or not

File: s/elements/bindings-editor/views/text-editor-panel.ts [hunk 6]
```diff
@@ -2,22 +2,53 @@
...
+		try {
+			setDraft(
+				parseBindings(target.value)
+			)
+		}
+		catch (error) {
+			if (error instanceof Error)
+				setError(error)
+		}
+	}
```
Comment (chase-moskal) [inline s/elements/bindings-editor/views/text-editor-panel.ts:31]: whenever user changes the textarea input, we parse the bindings to save the draft, and record any errors

File: s/elements/bindings-editor/views/text-editor-panel.ts [hunk 7]
```diff
@@ -2,22 +2,53 @@
...
 	return html`
 		<div data-panel=text-editor>
-			<textarea>${bindingsToText(bindings)}</textarea>
+			<textarea @input=${onTextAreaChange}>${text}</textarea>
+
+			${error && html`
+				<div class=error>
+					${error.message}
+				</div>
+			`}
```
Comment (chase-moskal) [inline s/elements/bindings-editor/views/text-editor-panel.ts:43]: if there's an error, we complain to the user about it

File: s/elements/bindings-editor/views/text-editor-panel.ts [hunk 8]
```diff
@@ -2,22 +2,53 @@
...
+
+			<div class=buttons>
+				${((draft && !error) && html`
+					<button @click=${() => onClickSave(draft)}>
+						save
+					</button>
+				`) || null}
```
Comment (chase-moskal) [inline s/elements/bindings-editor/views/text-editor-panel.ts:50]: when there's a draft, without errors, the save button is available

File: s/elements/context/element.ts [hunk 9]
```diff
@@ -41,6 +41,9 @@ export const NubContext = element<NubContextProperties>({
 		use.state<Bindings>(
 			element => load()
 				?? element.defaultBindingsJson
+				?? (element["default-bindings"]
+					? parseBindings(element["default-bindings"])
+					: null)
```
Comment (chase-moskal) [inline s/elements/context/element.ts:46]: now we actually use the html `default-bindings` attribute

File: s/types.ts [hunk 10]
```diff
@@ -35,25 +35,6 @@ export namespace Nub {
...
-	/** mouse action bindings */
-	"🖱": ActionBinds
-
-	/** vector2 action bindings */
-	"🕹️": ActionBinds
-
-	/** key action bindings */
-	"*️⃣": ActionBinds
-}
-
```
Comment (chase-moskal) [inline s/types.ts:52]: these types were moved to the new bindings area

**Review by chase-moskal — APPROVED** [review]
here's my work on improving the parser. it passes all the tests.

File: s/bindings/types.ts [hunk 11]
```diff
@@ -4,16 +4,8 @@ export interface Binds {
...
-
-	/** mouse action bindings */
 	mouse: Binds
-
-	/** vector2 action bindings */
 	vector2: Binds
-
-	/** key action bindings */
 	key: Binds
 }
```
Comment (chase-moskal) [inline s/bindings/types.ts:11]: removed these old comments. they were more useful when they were explaining the meaning of the emojis

File: s/elements/bindings-editor/views/text-editor-panel.ts [hunk 12]
```diff
@@ -14,19 +14,21 @@ export const TextEditorPanelView = view(use => ({
...
 	}) => {
 
-	const [error, setError] = use.state<undefined | Error>(undefined)
+	const [problem, setProblem] = use.state<string>("")
 	const [draft, setDraft] = use.state<undefined | Bindings>(undefined)
 
 	function onTextAreaChange(event: InputEvent) {
 		const target = <HTMLTextAreaElement>event.target
+		setProblem("")
+		setDraft(undefined)
```
Comment (chase-moskal) [inline s/elements/bindings-editor/views/text-editor-panel.ts:23]: this fixes the behavior of the panel

File: s/elements/bindings-editor/views/text-editor-panel.ts [hunk 13]
```diff
@@ -14,19 +14,21 @@ export const TextEditorPanelView = view(use => ({
 		onClickSave: (draft: Bindings) => void
 	}) => {
 
-	const [error, setError] = use.state<undefined | Error>(undefined)
+	const [problem, setProblem] = use.state<string>("")
```
Comment (chase-moskal) [inline s/elements/bindings-editor/views/text-editor-panel.ts:17]: now we're just using an "problem string" instead of an actual error object

**Review by chase-moskal — APPROVED** [review]
added a fix and some improvements

File: s/bindings/parse.test.ts [hunk 14]
```diff
@@ -4,123 +4,92 @@ import {Suite, expect} from "cynic"
...
-			},
-		})
-
-		return {
-			key: testChannel("key"),
-			mouse: testChannel("mouse"),
-			vector2: testChannel("vector2"),
-		}
-	},
+	"action binds": runTheseTestsOnEachBindType(channel => ({
```
Comment (chase-moskal) [inline s/bindings/parse.test.ts:11]: i managed to separate out this unsightly plumbing,

now this main test file looks much more elegant

File: s/bindings/parse.test.ts [hunk 15]
```diff
@@ -4,123 +4,92 @@ import {Suite, expect} from "cynic"
...
+			expect(a).equals("Alpha") },
+
+		"two binds": async() => {
+			const emoji = emojis[channel]
+			const binds = <Binds>parseBindings(`
+				${emoji} Benevolent :: Alpha Bravo `)[channel]
+			const [a, b] = binds["Benevolent"] ?? []
+			expect(a).equals("Alpha")
+			expect(b).equals("Bravo") },
+
```
Comment (chase-moskal) [inline s/bindings/parse.test.ts:34]: i'm experimenting with this weird syntax,

where i'm trying to "make it read like python"

i call it "tucked" style, and i think i might like it for test suites i think...

what do you think?

**Review by chase-moskal — APPROVED** [review]
i cleaned up the test suite, and produced some experimental syntax

File: s/bindings/parse.test.ts (hunk 15 above)
Comment (zenkyuv) [reply s/bindings/parse.test.ts:34]: Whats the difference between those styles ? If you like it I will try do adapt whenever I will write any test

File: s/bindings/emojis.ts [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+	// 🖱
+	mouse: [
+		128433, // U+1F5B1 three button mouse
+	],
+
+	// 🕹️
+	vector2: [
+		128377, // U+1F579 joystick
+	],
+} as const
```
[the remark this answers is not in the document]
Comment (zenkyuv) [reply s/bindings/emojis.ts:27]: yeah haha but emojis are what makes it fancy

File: s/bindings/parse.ts [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+			bindings,
+			match,
+		)
+	}
+
+	if (count === 0)
+		throw new Error("no valid statements found")
+
+	return bindings
+}
```
[the remark this answers is not in the document]
Comment (zenkyuv) [reply s/bindings/parse.ts:26]: it looks so elegant compared to my .. cant even call it code :weary::laughing:

File: s/elements/bindings-editor/element.ts (hunk 4 above)
Comment (zenkyuv) [reply s/elements/bindings-editor/element.ts:44]: where is that updateBindings ?

Comment (zenkyuv) [reply s/elements/bindings-editor/element.ts:44]: I see its in the context element, but how does it update bindings ?

**Comment by zenkyuv** [comment]
What can I say, this looks beautiful

**Pull request merged**